staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go

```diff
@@ -340,7 +340,7 @@ func transformResponseObject(ctx context.Context, scope *RequestScope, req *http
 		return
 	}
 	kind, serializer, _ := targetEncodingForTransform(scope, mediaType, req)
-	responsewriters.WriteObjectNegotiated(serializer, scope, kind.GroupVersion(), w, req, statusCode, obj, false)
+	responsewriters.WriteObjectNegotiated(serializer, scope, kind.GroupVersion(), w, req.WithContext(ctx), statusCode, obj, false)
```

Comment:
Looks like the context problem originates here?
https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go#L171-L173

Can we also attach the trace context to the request there? This would ensure we don't have a gap between that code and this code where the trace is not attached to the request.

Reply:
Thank you for taking the time to comment! :)
I understand you suggest we improve trace coverage to avoid gap as much as possible.
Could you please explain in more detail?

For below code, blocks `result, err := r.List(ctx, &opts)` 
and `transformResponseObject(ctx, scope, req, w, http.StatusOK, outputMediaType, result)` 
are already covered by parent span 
`ctx, span := tracing.Start(ctx, "List", traceFields(req)...)`.  

What else block would you suggest us to cover? 

staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go
```golang
func ListResource(r rest.Lister, rw rest.Watcher, scope *RequestScope, forceWatch bool, minRequestTimeout time.Duration) http.HandlerFunc {
	return func(w http.ResponseWriter, req *http.Request) {
		ctx := req.Context()
		// For performance tracking purposes.

                 // CasperLiu: this span is for high level - "List"
		ctx, span := tracing.Start(ctx, "List", traceFields(req)...)

		namespace, err := scope.Namer.Namespace(req)
		if err != nil {
			scope.err(err, w, req)
			return
		}

                 // omit...
		span.AddEvent("About to List from storage")
                 // CasperLiu: span inside below function is for low level - "List(recursive=true) etcd3"
		result, err := r.List(ctx, &opts)
		if err != nil {
			scope.err(err, w, req)
			return
		}
		span.AddEvent("Listing from storage done")
		defer span.AddEvent("Writing http response done", attribute.Int("count", meta.LenList(result)))
                 // CasperLiu: span inside below function is for low level - "SerializeObject"
		transformResponseObject(ctx, scope, req, w, http.StatusOK, outputMediaType, result)
	}
}
```

Here is the trace flow for the code mentioned above.
![image](https://github.com/kubernetes/kubernetes/assets/6135084/8a1149e0-9450-4bbc-a614-9361ace924ee)

Reply:
In get.go, instead of:

```go
 ctx := req.Context() 
 // For performance tracking purposes. 
 ctx, span := tracing.Start(ctx, "List", traceFields(req)...) 
```

could we do:

```go
 ctx := req.Context() 
 // For performance tracking purposes. 
 ctx, span := tracing.Start(ctx, "List", traceFields(req)...) 
req = req.WithContext(ctx)
```

?

My thinking here is:  If this is where we create the context that should be associated with the request for this code path, shouldn't set just set it up right here and that way if anyone uses the request after this point (in future edits of this code), they get a request with the correct context.

Reply:
+1, directly setting the request context right after creation makes it very clear

Reply:
Good suggestion! I would modify per your comment ~